include/GafferScene/SceneAlgo.inl

```diff
@@ -66,43 +67,41 @@ class TraverseTask : public tbb::task
 		{
 			ScenePlug::PathScope pathScope( m_context, m_path );
 
-			if( m_f( m_scene, m_path ) )
+			if( !m_f( m_scene, m_path ) )
 			{
-				IECore::ConstInternedStringVectorDataPtr childNamesData = m_scene->childNamesPlug()->getValue();
-				const std::vector<IECore::InternedString> &childNames = childNamesData->readable();
-
-				set_ref_count( 1 + childNames.size() );
-
-				ScenePlug::ScenePath childPath = m_path;
-				childPath.push_back( IECore::InternedString() ); // space for the child name
-				for( std::vector<IECore::InternedString>::const_iterator it = childNames.begin(), eIt = childNames.end(); it != eIt; it++ )
-				{
-					childPath[m_path.size()] = *it;
-					TraverseTask *t = new( allocate_child() ) TraverseTask( *this, childPath );
-					spawn( *t );
-				}
-				wait_for_all();
+				return NULL;
 			}
 
-			return NULL;
-		}
+			IECore::ConstInternedStringVectorDataPtr childNamesData = m_scene->childNamesPlug()->getValue();
+			const std::vector<IECore::InternedString> &childNames = childNamesData->readable();
+			if( childNames.empty() )
+			{
+				return NULL;
+			}
 
-	protected :
+			std::vector<ThreadableFunctor> childFunctors( childNames.size(), m_f );
 
-		TraverseTask( const TraverseTask &other, const ScenePlug::ScenePath &path )
-			:	m_scene( other.m_scene ),
-				m_context( other.m_context ),
-				m_f( other.m_f ),
-				m_path( path )
-		{
+			set_ref_count( 1 + childNames.size() );
+
+			ScenePlug::ScenePath childPath = m_path;
+			childPath.push_back( IECore::InternedString() ); // space for the child name
+			for( size_t i = 0, e = childNames.size(); i < e; ++i )
+			{
+				childPath.back() = childNames[i];
+				TraverseTask *t = new( allocate_child() ) TraverseTask( m_scene, m_context, childPath, childFunctors[i] );
+				spawn( *t );
+			}
+			wait_for_all();
+
```

Comment:
I should have made a todo comment for this, but I was thinking it would be really useful to call a method on the parent functor here, passing it the vector of child functors. This way it could gather results from children and percolate them back up to the root. For instance, this would allow matchingPaths to be implemented without any locks at all - each child could compute the PathMatcher relative to its location, and then the parent could combine them into a PathMatcher at its location and so on up the tree.

I guess we could introduce this later by using some template magic to call the method if it exists and not if it doesn't? Maybe we'd want that magic anyway, to keep things simple for the simple case.